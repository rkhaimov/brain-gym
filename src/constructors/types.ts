import { Either } from '../either';

export type TypeConstructor<TLeft, TRight> = (
  value: unknown
) => Either<TLeft, TRight>;
