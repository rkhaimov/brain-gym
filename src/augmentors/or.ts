import { TypeNode } from '../core';
import { validate } from '../validate';
import { refine } from './refine';
import { unknown } from '../types/unknown';

export const or = <TLeft, TRight>(
  left: TypeNode<TLeft>,
  right: TypeNode<TRight>
): TypeNode<TLeft | TRight> =>
  refine(unknown(), {
    validate: (value) => {
      const lerrors = validate(left, value as TLeft);

      if (lerrors.length === 0) {
        return [];
      }

      const rerrors = validate(right, value as TRight);

      if (rerrors.length === 0) {
        return [];
      }

      return [...lerrors, ...rerrors];
    },
    defaults: left.defaults,
  });
