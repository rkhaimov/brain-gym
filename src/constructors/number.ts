import { TypeConstructor } from './types';
import { Either } from '../either';

export const number = (): TypeConstructor<string, number> => (value) => {
  if (typeof value === 'number') {
    return Either.right(value);
  }

  return Either.left('is not a number');
};

