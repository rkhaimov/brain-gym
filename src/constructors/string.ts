import { TypeConstructor } from './types';
import { Either } from '../either';

export const string = (): TypeConstructor<string, string> => (value) => {
  if (typeof value === 'string') {
    return Either.right(value);
  }

  return Either.left('is not a string');
};
