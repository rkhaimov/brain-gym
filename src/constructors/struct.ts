import { TypeConstructor } from './types';
import { Either } from '../either';

export const struct =
  <TStruct extends Struct>(sct: TStruct): StructTypeConstructor<TStruct> =>
  (_value) => {
    if (typeof _value !== 'object') {
      return Either.left('is not an object');
    }

    if (_value === null) {
      return Either.left('is null');
    }

    const validations = Object.entries(sct).map(([key, property]) =>
      toKeyErrorPair(property(get(_value, key)), key)
    );

    return Either.sequenceA(validations).leftMap(
      toErrorStruct
    ) as StructEither<TStruct>;
  };

const get = (object: object, key: string): unknown =>
  (object as Record<string, unknown>)[key];

const toKeyErrorPair = <TLeft, TRight>(
  either: Either<TLeft, TRight>,
  key: string
): Either<[string, TLeft], TRight> => {
  return either.leftMap((left) => [key, left]);
};

const toErrorStruct = <TLeft>(
  errors: [string, TLeft][]
): Record<string, TLeft> => {
  return Object.fromEntries(errors);
};

type Struct = Record<string, TypeConstructor<unknown, unknown>>;

type StructTypeConstructor<TStruct> = TypeConstructor<
  StructToError<TStruct> | string,
  StructToType<TStruct>
>;

type StructEither<TStruct> = Either<
  StructToError<TStruct> | string,
  StructToType<TStruct>
>;

type StructToType<TStruct> = TStruct extends TypeConstructor<
  unknown,
  infer RType
>
  ? RType
  : { [P in keyof TStruct]: StructToType<TStruct[P]> };

type StructToError<TStruct> = TStruct extends TypeConstructor<
  infer RError,
  unknown
>
  ? RError
  : { [P in keyof TStruct]: StructToError<TStruct[P]> };
