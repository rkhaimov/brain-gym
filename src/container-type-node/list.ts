import { ToType, TypeNode } from '../core';
import { fromErrorMessage, prependErrorPathWith, validate } from '../validate';
import {
  ContainerTypeNode,
  createContainerTypeNode,
} from './createContainerTypeNode';

export const list = <TChildren extends TypeNode>(
  tn: TChildren
): ContainerTypeNode<ToType<TChildren>[], TChildren> =>
  createContainerTypeNode({
    validate: (value) => {
      if (Array.isArray(value)) {
        return value.flatMap((element, index) =>
          validate(tn, element).map(prependErrorPathWith(index))
        );
      }

      return [fromErrorMessage('Not a list')];
    },
    defaults: () => [],
    children: () => tn,
  });
