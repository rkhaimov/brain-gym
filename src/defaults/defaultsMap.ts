import { ToType, TypeNode, TypeNodeOperator } from '../core';

export const defaultsMap =
  <TTypeNode extends TypeNode>(
    transform: (defaults: ToType<TTypeNode>, tn: TTypeNode) => ToType<TTypeNode>
  ): TypeNodeOperator<TTypeNode> =>
  (tn) =>
    tn.__clone({
      defaults: () => transform(tn.defaults() as ToType<TTypeNode>, tn),
    });
