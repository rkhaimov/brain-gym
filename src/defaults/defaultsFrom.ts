import { ToType, TypeNode, TypeNodeOperator } from '../core';
import { defaultsMap } from './defaultsMap';
import { defaults } from './index';

export const defaultsFrom =
  <TTypeNode extends TypeNode>(
    from: TypeNode<ToType<TTypeNode>>
  ): TypeNodeOperator<TTypeNode> =>
  (tn) =>
    tn.wrap(defaultsMap(() => defaults(from)));
