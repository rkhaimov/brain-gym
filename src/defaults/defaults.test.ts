import { list } from '../types/list';
import { lengthMin } from '../operators/lengthMin';
import { or } from '../hot/or';
import { struct } from '../types/struct';
import { brand } from '../hot/brand';
import { moreThan } from '../operators/moreThan';
import { lessThan } from '../operators/lessThan';
import { boolean } from '../types/boolean';
import { nil } from '../types/nil';
import { number } from '../types/number';
import { string } from '../types/string';
import { defaults } from './index';
import { defaultsMap } from '../operators/defaultsMap';
import { defaultsFrom } from '../operators/defaultsFrom';
import { func } from '../types/func';
import { validate } from '../validate';

test('defaults of primitive values follows condition', () => {
  const tn = number();

  expect(defaults(tn)).toMatchInlineSnapshot(`0`);
});

test('defaults of primitives can be overridden', () => {
  const numeric = 10;
  const tn = number().pipe(defaultsMap(() => numeric));

  expect(defaults(tn)).toEqual(numeric);
});

test('defaults can be changed by operator to pass additional checks', () => {
  const tn = number().pipe(moreThan(50));

  expect(defaults(tn)).toMatchInlineSnapshot(`51`);
});

test('error is thrown when defaults do not pass own rules', () => {
  const tn = number().pipe(moreThan(5), lessThan(3));

  expect(() => defaults(tn)).toThrowErrorMatchingSnapshot();
});

test('manual defaults can be provided to avoid such errors', () => {
  const tn = number().pipe(
    moreThan(10),
    moreThan(5),
    defaultsMap(() => 20)
  );

  expect(defaults(tn)).toMatchInlineSnapshot(`20`);
});

test('or generates defaults from first type node', () => {
  const left = number();
  const right = nil();
  const tn = or(left, right);

  expect(defaults(tn)).toMatchInlineSnapshot(`0`);

  expect(defaults(tn.pipe(defaultsFrom(right)))).toMatchInlineSnapshot(
    `undefined`
  );
});

test('type augmentations support defaults', () => {
  const tn = brand(number(), 'id');

  expect(defaults(tn)).toMatchInlineSnapshot(`0`);
});

test('containers generates empty versions by default but it can be customized', () => {
  const tn = list(number());

  expect(defaults(tn)).toMatchInlineSnapshot(`Array []`);

  const rtn = tn.pipe(lengthMin(1));

  expect(() => defaults(rtn)).toThrowErrorMatchingInlineSnapshot(
    `"lengthMin: 1"`
  );

  const rrtn = rtn.pipe(defaultsMap(() => [defaults(tn.children())]));

  expect(defaults(rrtn)).toMatchInlineSnapshot(`
    Array [
      0,
    ]
  `);
});

test('struct generates defaults for each of its children', () => {
  const user = struct({
    name: string(),
    age: number().pipe(moreThan(10)),
    preferences: struct({ colors: list(string()), dark: boolean() }),
  });

  expect(defaults(user)).toMatchInlineSnapshot(`
    Object {
      "age": 11,
      "name": "",
      "preferences": Object {
        "colors": Array [],
        "dark": false,
      },
    }
  `);
});

test('func can generate default value', () => {
  const tn = func([number()], number());
  const generated = defaults(tn);

  expect(validate(tn, generated)).toMatchInlineSnapshot(`Array []`);
  expect(generated(2)).toMatchInlineSnapshot(`0`);
});
