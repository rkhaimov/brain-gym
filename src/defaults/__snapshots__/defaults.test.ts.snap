// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error is thrown when defaults do not pass own rules 1`] = `"moreThan: 5"`;
